pygeos/strtree.py

```diff
@@ -0,0 +1,24 @@
+from pygeos import lib
+import numpy as np
+
+__all__ = ["STRtree"]
+
+
+class STRtree:
+    """A query-only R-tree created using the Sort-Tile-Recursive (STR)
+    algorithm.
+
+    For two-dimensional spatial data.
+    """
+
+    def __init__(self, geometries, leafsize=5):
+        self._tree = lib.STRtree(np.asarray(geometries, dtype=np.object), leafsize)
+
+    def query(self, envelope):
+        """Return all items whose extent intersect the given envelope.
+
+        Parameters
+        ----------
+        envelope : Geometry
+        """
+        return self._tree.query(envelope)
```

Comment:
In shapely's version of STRTree, this is also the design to return the actual geometries. 
But for me, that has almost always been a reason to *not* use it (compared to rtree), as in most practical applications I needed indexes (of course because I was specifically working with dataframes, and not just geometries, so in those cases you often want to get the row of the returned geometry eventually, and then you need an index)

Reply:
Returning indices instead of the geometries itself seems to be a better approach indeed. That seems to be how scipy's KDTree does it:
https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.spatial.KDTree.html

Then we would have to store the original 1D ndarray on the STRtree object, dereference that on deallocation (which would in turn dereference the contained geometries).

A big downside would be that we would have memory leaks / segfaults if one changes the array after constructing the STRtree. If we do this, we should put an explicit warning that this array should not be changed in the docs. By the way, also how scipy does it: "This array is not copied, and so modifying this data will result in bogus results."